EpiAware/src/EpiInfModels/InfectionODEProcess.jl

```diff
@@ -0,0 +1,182 @@
+    u0::VecOrMat{T}
+    p::VecOrMat{T}
+end
+
+function ODEParams(; u0::VecOrMat, p::VecOrMat)
+    T = promote_type(eltype(u0), eltype(p))
+    return ODEParams{T}(T.(u0), T.(p))
+end
+
+"""
+A structure representing an infection process modeled by an Ordinary Differential Equation (ODE).
+
+# Background
+
+The purpose of this structure is to define the behaviour of modelling an infection process
+using an ODE. We use the [`SciML` ecosystem](https://sciml.ai/) to define and solve the ODE.
+For `InfectionODEProcess` structs we focus on defining from a restricted set of ODE problems:
+- The initial condition `u0` must be a vector or matrix.
+- The parameters `p` must be a vector or matrix.
+- The output of the ODE should be interpreted as the infection incidence at each time point in
+`ts` via the function `sol2infs` which maps the solution object `sol` of the ODE solve to infection counts.
+
+# Fields
+- `prob::P`: The ODE problem instance, where `P` is a subtype of `ODEProblem`.
+- `ts::Vector{T}`: A vector of time points, where `T` is the type of the time points.
+- `solver::S`: The solver used for the ODE problem.
+- `sol2infs::F`: A function that maps the solution object of the ODE to infection counts.
+
+# Constructors
+- `InfectionODEProcess(prob::ODEProblem; ts, solver, sol2infs)`: Create an `InfectionODEProcess`
+object with the ODE problem `prob`, time points `ts`, solver `solver`, and function `sol2infs`.
+
+# Type Parameters
+- `P`: The type of the ODE problem, which must be a subtype of `ODEProblem`.
+- `T`: The type of the time points.
+- `S`: The type of the solver.
+- `F`: The type of the function that maps the solution to infection counts, which must be a subtype of `Function`.
+
+# Example
+
+```jldoctest
+using EpiAware, OrdinaryDiffEq
+r = log(2) / 7 # Growth rate corresponding to 7 day doubling time
+u0 = [1.0]
+p = [r]
+params = ODEParams(u0 = u0, p = p)
+
+# Define the ODE problem using SciML
+# We use a simple exponential growth model
+
+function expgrowth(du, u, p, t)
+    du[1] = p[1] * u[1]
+end
+prob = ODEProblem(expgrowth, u0, (0.0, 10.0), p)
+
+# Define the InfectionODEProcess
+
+expgrowth_model = InfectionODEProcess(prob::ODEProblem; ts = 0:1:10,
+        solver = Tsit5(),
+        sol2infs = sol -> sol[1, :])
+
+# Generate the latent infections
+I_t = generate_latent_infs(expgrowth_model, params)()
+
+# output
+
+11-element Vector{Float64}:
+ 1.0
+ 1.1040895124087677
+ 1.2190137467993492
+ 1.3459001375697022
+ 1.4859941865014936
+ 1.640671113705054
+ 1.8114471151863056
+ 1.9999990356297939
+ 2.2081789476865237
+ 2.438027196361022
+ 2.6918002758361723
+```
+"""
+struct InfectionODEProcess{P, T, S, F} <: EpiAwareBase.AbstractTuringEpiModel where {
+    P <: ODEProblem,
+    T,
+    S,
+    F <: Function}
+    prob::P
+    ts::Vector{T}
+    solver::S
+    sol2infs::F
+end
+
+function InfectionODEProcess(prob::ODEProblem; ts,
+        solver = AutoTsit5(Rosenbrock23()),
+        sol2infs = sol -> sol[end, :])
```

Comment:
maybe call in transform to match other args elsewhere that do a similar thing?

Reply:
I thought about this, but we have a few things going on:

- The concept of reduction (rather than transformation), e.g. going from an `ODESolution` struct to an `Array`.
- The concept of transforming that `Array` for the purpose of easier AD/ catching non-negs without needing to use `DiffEqCallbacks` etc
- The concept of transforming as a part of a model (as separate from convenience), e.g. our `Transform...` model structs.

So it felt that using `transform` willy-nilly might be even more confusing?

Reply:
fine can use something that doesn't include infs. (reduction, selection